src/main/java/org/zalando/nakadi/service/subscription/SubscriptionTimeLagService.java

```diff
@@ -0,0 +1,129 @@
+        this.timelineService = timelineService;
+        this.cursorComparator = cursorComparator;
+    }
+
+    public Map<EventTypePartition, Duration> getTimeLags(final Collection<NakadiCursor> committedPositions,
+                                                         final List<PartitionEndStatistics> endPositions)
+            throws ErrorGettingCursorTimeLagException, InconsistentStateException {
+
+        final ExecutorService executor = Executors.newFixedThreadPool(LAG_CALCULATION_PARALLELISM);
+
+        final Map<EventTypePartition, Future<Duration>> futures = new HashMap<>();
+        final Map<EventTypePartition, Duration> result = new HashMap<>();
+
+        for (final NakadiCursor cursor : committedPositions) {
+            final EventTypePartition partition = new EventTypePartition(cursor.getTimeline().getEventType(),
+                    cursor.getPartition());
+            if (isCursorAtTail(cursor, endPositions)) {
+                result.put(partition, Duration.ZERO);
+            } else {
+                futures.put(partition, executor.submit(() -> getNextEventTimeLag(cursor)));
+            }
+        }
+
+        executor.shutdown();
+        try {
+            final boolean finished = executor.awaitTermination(COMPLETE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            if (!finished) {
+                throw new InconsistentStateException("Timeout occurred when getting subscription time lag");
+            }
+            for (final EventTypePartition partition : futures.keySet()) {
+                final Future<Duration> future = futures.get(partition);
+                try {
+                    result.put(partition, future.get());
+                } catch (final ExecutionException e) {
+                    throw e.getCause();
+                }
+            }
+            return result;
+
+        } catch (final InconsistentStateException | ErrorGettingCursorTimeLagException e) {
+            throw e;
+        } catch (final Throwable e) {
+            throw new InconsistentStateException("Unexpected error occurred when getting subscription time lag", e);
+        }
+    }
+
+    private Duration getNextEventTimeLag(final NakadiCursor cursor) throws ErrorGettingCursorTimeLagException,
+            InconsistentStateException {
+        try {
+            final EventConsumer consumer = timelineService.createEventConsumer(
+                    "time-lag-checker-" + UUID.randomUUID().toString(), ImmutableList.of(cursor));
+
+            final ConsumedEvent nextEvent = executeWithRetry(
+                    () -> {
+                        final List<ConsumedEvent> events = consumer.readEvents();
+                        return events.isEmpty() ? null : events.iterator().next();
+                    },
+                    new RetryForSpecifiedTimeStrategy<ConsumedEvent>(EVENT_FETCH_WAIT_TIME_MS)
+                            .withResultsThatForceRetry((ConsumedEvent) null));
+
+            if (nextEvent == null) {
```

Comment:
Shouldn't this be an error condition? If I understood correctly, before invoking this method you check if the cursor is not the END. So there should always be a "next". Then this should throw an error because we were not able to fetch the event on time.

Reply:
I agree. Fixed https://github.com/zalando/nakadi/pull/873/commits/36ff06ad6adb6c505d566d59419174244969d7e7